ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerImpl.java

```diff
@@ -0,0 +1,170 @@
+        )
+    );
+  }
+
+  @VisibleForTesting
+  TransactionalProducerImpl(
+      final String commandTopicName,
+      final CommandRunner commandRunner,
+      final Duration commandQueueCatchupTimeout,
+      final Consumer<CommandId, Command> commandConsumer,
+      final Producer<CommandId, Command> commandProducer
+  ) {
+    this.commandTopicPartition = new TopicPartition(
+        Objects.requireNonNull(commandTopicName, "commandTopicName"),
+        0
+    );
+    this.commandQueueCatchupTimeout = 
+        Objects.requireNonNull(commandQueueCatchupTimeout, "commandQueueCatchupTimeout");
+    this.commandConsumer = Objects.requireNonNull(commandConsumer, "commandConsumer");
+    this.commandProducer = Objects.requireNonNull(commandProducer, "commandProducer");
+    this.commandTopicName = Objects.requireNonNull(commandTopicName, "commandTopicName");
+    this.commandRunner = Objects.requireNonNull(commandRunner, "commandRunner");
+    initialize();
+  }
+
+  private void initialize() {
+    commandConsumer.assign(Collections.singleton(commandTopicPartition));
+    commandProducer.initTransactions();
+  }
+
+  /** begins transaction */
+  public void begin() {
+    commandProducer.beginTransaction();
+    waitForConsumer();
+  }
+
+  public void waitForConsumer() {
+    try {
+      final long endOffset = getEndOffset();
+      
+      // timeout hardcoded for now
+      commandRunner.ensureProcessedPastOffset(endOffset - 1, commandQueueCatchupTimeout);
+    } catch (final InterruptedException e) {
+      final String errorMsg = 
+          "Interrupted while waiting for commandRunner to process command topic";
+      throw new KsqlRestException(
+          Errors.serverErrorForStatement(e, errorMsg, new KsqlEntityList()));
+    } catch (final TimeoutException e) {
+      final String errorMsg =
+          "Timeout while waiting for commandRunner to process command topic";
+      throw new KsqlRestException(
+          Errors.serverErrorForStatement(e, errorMsg, new KsqlEntityList()));
+    }
+  }
+
+  private long getEndOffset() {
+    return commandConsumer.endOffsets(Collections.singletonList(commandTopicPartition))
+        .get(commandTopicPartition);
+  }
+
+  public RecordMetadata send(final CommandId commandId, final Command command) {
+    final ProducerRecord<CommandId, Command> producerRecord = new ProducerRecord<>(
+        commandTopicName,
+        0,
+        Objects.requireNonNull(commandId, "commandId"),
+        Objects.requireNonNull(command, "command"));
+    try {
+      return commandProducer.send(producerRecord).get();
+    } catch (final ExecutionException e) {
+      if (e.getCause() instanceof RuntimeException) {
+        throw (RuntimeException) e.getCause();
+      }
+      throw new RuntimeException(e.getCause());
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    } catch (KafkaException e) {
+      commandProducer.abortTransaction();
```

Comment:
`begin()` is called in DistributingExecutor.execute()
The entire protocol for begin transaction -> validate statement -> send to command topic -> commit transaction happens in that function. Since it's all wrapped in a try catch, if any exception is thrown in DistributingExecutor.execute(), any existing transaction will be aborted.